mountpoint-s3/src/inode.rs

```diff
@@ -354,6 +354,65 @@ impl Superblock {
 
         Ok(LookedUp { inode, stat })
     }
+
+    /// Remove local-only empty directory, i.e., the ones created by mkdir.
+    /// It does not affects empty directories represented remotely with directory markers.  
+    pub async fn rmdir<OC: ObjectClient>(
+        &self,
+        client: &OC,
+        parent_ino: InodeNo,
+        name: &OsStr,
+    ) -> Result<(), InodeError> {
+        let LookedUp { inode, .. } = self.lookup(client, parent_ino, name).await?;
+
+        if inode.kind() == InodeKind::File {
+            return Err(InodeError::NotADirectory(inode.ino()));
+        }
+
+        let parent = self.inner.get(parent_ino)?;
+        let mut parent_state = parent.inner.sync.write().unwrap();
+
+        let mut inode_state = inode.inner.sync.write().unwrap();
+
+        match inode_state.write_status {
+            WriteStatus::Remote => {
+                error!(parent = parent_ino, ?name, "rmdir called on a remote directory",);
+                return Err(InodeError::RemoteDirectory(inode.ino()));
+            }
+            WriteStatus::Deleted => {
+                error!(parent = parent_ino, ?name, "rmdir called on a non existent directory",);
+                return Err(InodeError::InodeDeleted(inode.ino()));
+            }
+            // Although Local Open is not possible for directories, added it to be on safe side.
+            WriteStatus::LocalOpen | WriteStatus::LocalUnopened => {
+                let inode_kind_data = &inode_state.kind_data;
+                // A local directory can only contain local children.
+                if let InodeKindData::Directory { writing_children, .. } = inode_kind_data {
+                    if !writing_children.is_empty() {
+                        return Err(InodeError::DirectoryNotEmpty(inode.ino()));
+                    }
+                }
+            }
+        }
+
+        match &mut parent_state.kind_data {
+            InodeKindData::File {} => {
+                debug_assert!(false, "inodes never change kind");
+                return Err(InodeError::NotADirectory(parent.ino()));
+            }
+            InodeKindData::Directory {
+                children,
+                writing_children,
+            } => {
+                writing_children.remove(&inode.ino());
+                children.remove(inode.name());
```

Comment:
Should we assert something about these? If it's a local directory, it must have been in both these hashmaps, right?